packages/php-wasm/universal/src/lib/base-php.ts

```diff
@@ -101,10 +101,16 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
 			throw new Error('Invalid PHP runtime id.');
 		}
 		this[__private__dont__use] = runtime;
-		runtime['onMessage'] = (data: string) => {
+		runtime['onMessage'] = async (data: string): Promise<string> => {
 			for (const listener of this.#messageListeners) {
-				listener(data);
+				const returnData = await listener(data);
+
+				if (returnData) {
+					return returnData;
+				}
 			}
+
+			return '';
```

Comment:
@adamziel Maybe this should change too. Also currently it returns when any callback returns any data, not sure if it's an ok assumption that only one callback will return.